src/main/java/dev/pgm/community/sessions/services/SQLSessionService.java

```diff
@@ -0,0 +1,202 @@
+package dev.pgm.community.sessions.services;
+
+import com.google.common.cache.CacheBuilder;
+import com.google.common.cache.CacheLoader;
+import com.google.common.cache.LoadingCache;
+import dev.pgm.community.database.DatabaseConnection;
+import dev.pgm.community.feature.SQLFeatureBase;
+import dev.pgm.community.sessions.Session;
+import dev.pgm.community.sessions.SessionQuery;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.time.Instant;
+import java.util.List;
+import java.util.UUID;
+import java.util.concurrent.CompletableFuture;
+import tc.oc.pgm.util.concurrent.ThreadSafeConnection.Query;
+
+public class SQLSessionService extends SQLFeatureBase<Session, SessionQuery> {
+
+  private static final String TABLE_NAME = "sessions";
+  private static final String TABLE_FIELDS =
+      "(id VARCHAR(36) PRIMARY KEY, player VARCHAR(36), disguised BOOL, server VARCHAR(32), startTime LONG, endTime LONG)";
+
+  private LoadingCache<SessionQuery, SelectLatestSessionQuery> sessionCache;
+
+  public SQLSessionService(DatabaseConnection database) {
+    super(database, TABLE_NAME, TABLE_FIELDS);
+    this.sessionCache =
+        CacheBuilder.newBuilder()
+            .build(
+                new CacheLoader<SessionQuery, SelectLatestSessionQuery>() {
+                  @Override
+                  public SelectLatestSessionQuery load(SessionQuery key) throws Exception {
+                    return new SelectLatestSessionQuery(
+                        key.getPlayerId(), key.filterOutDisguised());
+                  }
+                });
+  }
+
+  @Override
+  public void save(Session session) {
+    SelectLatestSessionQuery query =
+        sessionCache.getIfPresent(new SessionQuery(session.getPlayerId(), false));
+    if (query != null) query.invalidate();
+
+    query = sessionCache.getIfPresent(new SessionQuery(session.getPlayerId(), true));
+    if (query != null) query.invalidate();
+
+    getDatabase().submitQuery(new InsertQuery(session));
+  }
+
+  public void updateSessionEndTime(Session session) {
+    getDatabase().submitQuery(new UpdateSessionEndTimeQuery(session));
+  }
+
+  @Override
+  public CompletableFuture<List<Session>> queryList(SessionQuery target) {
+    return null; // not implemented
+  }
+
+  @Override
+  public CompletableFuture<Session> query(SessionQuery target) {
+    SelectLatestSessionQuery query = sessionCache.getUnchecked(target);
+
+    if (query.hasFetched()) {
+      return CompletableFuture.completedFuture(query.getSession());
+    } else {
+      return getDatabase()
+          .submitQueryComplete(query)
+          .thenApplyAsync(q -> SelectLatestSessionQuery.class.cast(q).getSession());
+    }
+  }
+
+  private class InsertQuery implements Query {
```

Comment:
Idea warns that thise and the other queries inner classes can and should be static
```suggestion
  private static class InsertQuery implements Query {
```